File: code/dynamic_programming/factorial/factorial.scala
/* Part of Cosmos by OpenGenus Foundation */

import scala.annotation.tailrec

/**
  * Implement factorial using tail recursion
  */
object main {

  @tailrec
  private def _factorial(of: Int, current: Int): Int = {
    if (of == 0) {
      current
    } else {
      _factorial(of - 1, of * current)
    }
  }

  def factorial(of: Int): Int = {
    _factorial(of, 1)
  }

  def main(args: Array[String]) = {
    val number: Int = 5
    println(s"The factorial of ${number} is ${factorial(number)}")
  }
}

File: code/dynamic_programming/src/factorial/factorial.rs
// Part of Cosmos by OpenGenus Foundation
fn factorial(n: u64) -> u64 {
    match n {
        0 => 1,
        _ => n * factorial(n - 1)
    }
}

fn main() {
    for i in 1..10 {
        println!("factorial({}) = {}", i, factorial(i));
    }
}
